rfc/remove-ucan-stream.md

```diff
@@ -0,0 +1,55 @@
+* Tasks that depend on the receipt of a different invocation can exert influence over it's content. For example, in billing we discovered that storing the same blob multiple times was causing double counts. This is undesirable - a user should not be charged multiple times for the same item of data. At the time the _easiest_ solution was to alter the receipt the billing service relied upon to include the number of allocated bytes. Essentially if the blob already exists in the space then the allocated bytes are 0, but if the blob does not then the allocated bytes are equal to the size of the blob. It sounds sensible, but the change to the receipt is not needed by the `upload-api` sub-system at all. It is only relevant to the `billing` sub-system and worse, there is no way to know this is the case. It is a loose coupling between `billing` and `upload-api` sub-systems which is brittle for reasons stated above. A better solution might be to define specific usage invocations for billing and to invoke them at the appropriate point when an additon is known to affect space size.
+
+## Proposal
+
+In general the proposal is to transition these tasks from async to sync. This gives us better visibility over what tasks need to be performed and when, due to the fact that they are executed at the point they are required, not later. It makes explicit the coupling between sub-systems where it was previously loose and implicit. It also surfaces failures earlier and provided we ensure atomicity, will maintain system consistency.
+
+We SHOULD be careful to ensure atomic writes when an invocation writes to multiple places. In dynamodb this can easily be achieved using `TransactWriteItems`, or alternatively we can "rollback" succeeded tasks when others fail.
+
+The risk is that our invocations take a little longer to execute, but consistency within the system is more desirable than speed.
+
+The work in [Consolidating `w3up` + `upload-service`](https://github.com/storacha/RFC/blob/main/rfc/consolidating-w3up-and-upload-service.md) actually includes porting a change made to `upload-service-infra` that switches metrics recording to be done at the time events happen, rather than asynchronously later. This was done as an experimental effort towards the proposal in this RFC. Note: this does not include any change to metrics collection for Store Protocol invocations.
+
+### Short-term
+
+Switch the UCAN stream handler to use `web3.storage/blob/accept` receipts, which do not currently appear in another invocation response and thus are not duplicated on the stream.
+
+This re-exposes the bug where two uploads of the same item are double counted, but it is categorically better than double counting _every_ upload.
+
+The PR for this change can be found here: https://github.com/storacha/w3infra/pull/456
+
+### Mid-term
+
+1. Insert space usage deltas directly to DynamoDB table from the `BlobRegistry` at the point where items are added or removed.
```

Comment:
Just to confirm, the `BlobRegistry` will be available after https://github.com/storacha/w3infra/pull/453 is merged, right? Also, are the `register` and `deregister` methods only used in `blob/accept` and `blob/remove`, respectively?  
About the error handler, what’s currently being done with failed requests? Are they stored somewhere so we can reprocess them later? Are we satisfied with how things are, or should it also be discussed?

Reply:
> Just to confirm, the `BlobRegistry` will be available after [storacha/w3infra#453](https://github.com/storacha/w3infra/pull/453) is merged, right?

Correct

> Also, are the `register` and `deregister` methods only used in `blob/accept` and `blob/remove`, respectively?

Correct.

> About the error handler, what’s currently being done with failed requests? Are they stored somewhere so we can reprocess them later? Are we satisfied with how things are, or should it also be discussed?

What requests are you referring to?

Failed invocations are surfaced by the client, so it is up to the caller to retry. We use `p-retry` internally when it makes sense.